### ISSUE TITLE
Bug Fixes/Enhancements to Application Gateway

### ISSUE THREAD

**Issue by tombuildsstuff** [terraform-providers/terraform-provider-azurerm#1576]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

* `azurerm_application_gateway`

---

The `azurerm_application_gateway` resource is currently missing a selection of fields and also has some bugs which need resolving. Unfortunately there's [a bug in the Application Gateway API where the Application Gateway isn't actually deleted](https://github.com/Azure/azure-rest-api-specs/issues/2187) which prevents us from proceeding this work, since our tests for these resources are failing around 80% of the time due to the Application Gateway not being deleted (meaning this fills up our quota's).

Rather than trying to track the status of this bug across multiple issues - I'm opening this meta-issue to keep track of these bugs and enhancements in one place. Once the bug in the API is resolved - it should be possible to add these enhancements/investigate fixing these bugs; however these are blocked at the moment.

## Blocking API issues

- ~[The Application Gateway API returns that an Application Gateway has been deleted when it hasn't](https://github.com/Azure/azure-rest-api-specs/issues/2187) (which was previously tracked in this repository in https://github.com/terraform-providers/terraform-provider-azurerm/issues/608)~

## Enhancements

 - ~[Support for Disabled Rule Groups](https://github.com/terraform-providers/terraform-provider-azurerm/issues/451)~ (fixed in #3394)
 - ~[Support for Redirect Rules](https://github.com/terraform-providers/terraform-provider-azurerm/issues/552)~
 - ~[Support for Affinity cookie name](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1519)~ (fixed in #3434)
 - ~[Support for Connection draining](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1519)~ (fixed in #2778)
 - ~[Support for Hostname](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1875)~ (fixed in #2990)
- ~[Support for Diagnostics Logs](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1519)~ (this is the regular Diagnostics resource, being tracked in #657)
- ~[Split Application Gateway resource into multiple resources](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1542)~ not possible at this time due to the API
- ~[Tags](
https://github.com/terraform-providers/terraform-provider-azurerm/issues/1576#issuecomment-406302735)~ (fixed in #2054)

## Bug Fixes

 - ~[SSL Certificate recreated every apply](https://github.com/terraform-providers/terraform-provider-azurerm/issues/583)~ (fixed in #2054)
 - ~[Changing the Subnet fails](https://github.com/terraform-providers/terraform-provider-azurerm/issues/747)~ (fixed in #3437)
 - [Virtual Network Peering gets disconnected when changing a backend pool](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1274)

**Comment by mb290** [comment]
@tombuildsstuff can I request the support for tags be introduced with this too? https://docs.microsoft.com/en-us/rest/api/application-gateway/applicationgateways/updatetags

**Comment by tombuildsstuff** [comment]
@mb290 sure, I've updated the list :)

**Comment by guitarrapc** [comment]
@thomastaylor312 can I request the support for key vault certificate integration for https pfx?
It would be great if I can use with vault id reference like azurerm_virtual_machine resource.

> https://www.terraform.io/docs/providers/azurerm/r/virtual_machine.html#source_vault_id

**Comment by cdhunt** [comment]
Would this work include support for custom [SslPolicy](https://docs.microsoft.com/en-us/rest/api/application-gateway/applicationgateways/createorupdate#applicationgatewaysslpolicy) as mentioned in #1536? I don't see specific mention of it in the referenced issues.

**Comment by thomastaylor312** [comment]
@guitarrapc Did you mean to tag me on this?

**Comment by guitarrapc** [comment]
@thomastaylor312  Oh, I'm sorry, I was misspelled. I meant @tombuildsstuff.

**Comment by mb290** [comment]
@tombuildsstuff is this something actively being worked on now that the MS issue has been resolved? Aware that you're wanting to split up into multiple resources, I don't want to start doing PRs for any of the functionality I (selfishly) need if it's going to trip up over other people.

**Comment by tombuildsstuff** [comment]
@mb290 

> is this something actively being worked on now that the MS issue has been resolved? Aware that you're wanting to split up into multiple resources, I don't want to start doing PRs for any of the functionality I (selfishly) need if it's going to trip up over other people.

We're not actively working on this at this time - but we may in the future now this is unblocked. If you're interested in sending a PR for some of this functionality (thanks in advance 😄) I'd suggest it's probably worth mentioning here / in the relevant GH issue so other folks are aware / don't start working on the same thing (and we can update the description above)?

It's worth noting that the API work required to allow splitting this into multiple resources hasn't been completed yet, so I don't believe we're able to proceed with that due to the validation requirements at the API end) - but I may have mis-interpreted which task you were referring to?

Thanks!

**Comment by mb290** [comment]
@tombuildsstuff the two bits of functionality i'm wanting to implement are:

Tags - looks like this is now working with the recent MS fix so just a documentation update.
Documentation - the example actually has multiple different configurations all lumped in the same (with some syntax errors) so will tidy this up.
SSL Ciphers & Protocol Versions - linking to [#1536](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1536), [#451](https://github.com/terraform-providers/terraform-provider-azurerm/issues/451) & [#898](https://github.com/terraform-providers/terraform-provider-azurerm/issues/898)

Certainly not planning on undertaking the full split out of the provider - lets keep walking before I can run 😄.

**Comment by Supermathie** [comment]
@tombuildsstuff to ensure it doesn't get lost, can you add "Support for setting Diagnostic Logs" from https://github.com/terraform-providers/terraform-provider-azurerm/issues/1519 to the OP?

**Comment by tombuildsstuff** [comment]
@supermathie done :)

**Comment by steve-88** [comment]
Hi There, 
Any update on implementing AGW Redirects to the Routing Rules?

Thanks

**Comment by steve-88** [comment]
Another chase here. Maybe @tombuildsstuff ? I would love to get his working.

**Comment by tombuildsstuff** [comment]
@steve-88 the enhancements to Application Gateway aren't something I don't expect us to get to it anytime soon - however we'd happily accept a community contribution for it :)

**Comment by RustyF** [comment]
I too need the redirect stuff.  @tombuildsstuff - what's the learning curve like for this stuff (for an experienced .Net dev but Go newbie)?

**Comment by tombuildsstuff** [comment]
hey @RustyF

> I too need the redirect stuff. @tombuildsstuff - what's the learning curve like for this stuff (for an experienced .Net dev but Go newbie)?

So there's a few changes required but it should be fairly familiar - once you've got Go installed you'd need to make the following changes:

1. Creating the schema for Redirect Rules e.g. https://github.com/terraform-providers/terraform-provider-azurerm/blob/c9fb7f220b16f4df0e6cf61700f6c30fbcb9472e/azurerm/resource_arm_application_gateway.go#L492-L559
2. Adding the property to the Create/Update method: https://github.com/terraform-providers/terraform-provider-azurerm/blob/c9fb7f220b16f4df0e6cf61700f6c30fbcb9472e/azurerm/resource_arm_application_gateway.go#L736
3. Setting the value back in the Read function: https://github.com/terraform-providers/terraform-provider-azurerm/blob/c9fb7f220b16f4df0e6cf61700f6c30fbcb9472e/azurerm/resource_arm_application_gateway.go#L818-L822
4. Adding an acceptance test to cover this: https://github.com/terraform-providers/terraform-provider-azurerm/blob/c9fb7f220b16f4df0e6cf61700f6c30fbcb9472e/azurerm/resource_arm_application_gateway_test.go#L175-L199
5. Adding some documentation for the new field: https://github.com/terraform-providers/terraform-provider-azurerm/blame/c9fb7f220b16f4df0e6cf61700f6c30fbcb9472e/website/docs/r/application_gateway.html.markdown#L264

There's [details in the README](https://github.com/terraform-providers/terraform-provider-azurerm#developing-the-provider) about how to build the provider locally / run the tests - which should then be possible via:

```
TF_ACC=1 go test ./azurerm -v -timeout 120m -run=TestNameOrPrefixToRun -count=1
```

Hope that helps :)

**Comment by hugree** [comment]
Redirect rules are available in Azure Application Gateway for like a year and there are still no updates in terraform?!? I'm disappointed.

I used a null_resource with local_provisioner but that's a workaround and my deploy takes longer now.

**Comment by hdost** [comment]
@tombuildsstuff  Since this is being used as a major tracking issue I am actually interested in creating the resources in #1893 are the namings of all of those resources good? Should we discuss that here?

**Comment by steve-88** [comment]
Not only are they not interested in adding it...when you configure it manually, terraform has the cheek to revert it
  ~ azurerm_application_gateway.agw_web
      request_routing_rule.4.backend_address_pool_name:  "" => "xx-web-prod-agw-bepool"
      request_routing_rule.4.backend_http_settings_name: "" => "www.xxxxx.com-http"

**Comment by hdost** [comment]
@steve-88 Out of curiosity, when you say "they" do you mean the project maintainers? And that they aren't interested in adding it, would they accept a patch?

**Comment by hdost** [comment]
@steve-88 based on this and the comment above https://github.com/terraform-providers/terraform-provider-azurerm/issues/1576#issuecomment-422938664

They are more than willing to accept a patch. So I wouldn't say they aren't interested in adding it, but that they have other priorities. Maintainers have lots of requests they can't do them all at once.

**Comment by seitosan** [comment]
Hello Guys,

Have you already implemented pick hostname from backend ? 

best regards

**Comment by tombuildsstuff** [comment]
@seitosan support for this is being tracked in #1875

**Comment by tombuildsstuff** [comment]
> @tombuildsstuff Since this is being used as a major tracking issue I am actually interested in creating the resources in #1893 are the namings of all of those resources good? Should we discuss that here?

@hdost the API doesn't support this at this time - so I don't think we can plan this at this time, unfortunately, incase things change.

**Comment by dopsy** [comment]
Guys, due to terraform not having this implemented yet, is there a way to avoid terraform overwriting my rules for http>https redirection which i configure using local-provisioner? Thanks!

    provisioner "local-exec" {
        command     = "az network application-gateway redirect-config create --gateway-name ${azurerm_application_gateway.aag.name} -g ${var.rgs_name} -n ${var.prefix}-http-to-https --type Permanent --include-path true --include-query-string true --target-listener ****-listener"
    }

    provisioner "local-exec" {
        command     = "az network application-gateway rule create --gateway-name ${azurerm_application_gateway.aag.name} -g ${var.rgs_name} -n ${var.prefix}-Rule-http-https --rule-type Basic --http-listener ***http-listener --redirect-config ${var.prefix}-http-to-https"
    }
    lifecycle {
      ignore_changes = ["provisioner"]
    }


  ~ azurerm_application_gateway.aag
      request_routing_rule.#:                        "2" => "1"
      request_routing_rule.1.http_listener_name:     "****-listener" => ""
      request_routing_rule.1.name:                   "***-Rule-http-https" => ""
      request_routing_rule.1.rule_type:              "Basic" => ""

**Comment by seitosan** [comment]
@dopsy  Yes i used local-exec to customize the pickhostname. I supposed you can used it to overwriting your rules.

I have an issue with this kind of script, if you redeploy your application, the terraform will rollback your script and erase your local-exec. I haven't found the way to persist this configuration.

**Comment by dopsy** [comment]
For now i have found a workaround, which might be useful for someone else.

**Comment by tombuildsstuff** [comment]
👋

Since this is a meta-issue containing a few different issues, to try and better keep track of what's outstanding in this issue I've gone through and updated comments inline to mark what's fixed (and hidden some comments).

#2054 brings a bunch of fixes to the Application Gateway resource (some of which are outlined below) - once that's merged it should be possible to add the functionality outlined above.

- documentation's been rewritten
- ssl certificates/passwords no longer continually diff
- tags support (it was technically already present, but we've documented it now 😄)
- and a tonof bug fixes for things that weren't being set into the state separately

Thanks!

**Comment by mhennecke** [comment]
The description mentions #1519 is fixed by #2054. However, #1519 also requests the probe setting "pickHostNameFromBackendHttpSettings". Should I create a new feature request for that or should this be added to this meta-issue?

**Comment by kansberry** [comment]
I too would like an update on the redirect rules for the application gateway [#552](https://github.com/terraform-providers/terraform-provider-azurerm/issues/552). Is there a timeline for this implementation, or did @RustyF make any progress?

**Comment by steve-88** [comment]
It would indeed be awesome to get these AGW redirect requests sorted

**Comment by tombuildsstuff** [comment]
@Moeser @kansberry @steve-88 unfortunately we don't have a timeframe for us implementing this, but we'd happily accept a community contribution for this functionality if someone's interested in adding support for this in the interim - there's some more information here: https://github.com/terraform-providers/terraform-provider-azurerm/issues/1576#issuecomment-422938664

**Comment by RustyF** [comment]
I'm still alive :) and (just) making a start on it.

**Comment by RustyF** [comment]
Hi Tom

 

What branch are those changes from or is it a defunct branch now as seems to be a free-floating commit?

 

Thanks in advance

Russ

**Comment by tombuildsstuff** [comment]
@RustyF [master](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_application_gateway.go#L324-L391) - it's just that file's been refactored since then :)

**Comment by RustyF** [comment]
There's not a lot to do but it's just taking me ages as I'm learning Go and grokking the overall design at the same time.

**Comment by mcharriere** [comment]
Hi, I've been working on adding support for Custom errors to the application gateway resource. @tombuildsstuff  would you like to add this new feature to the enhancements list? 

Should I open a pull request? I have the code in this [repo](https://github.com/mcharriere/terraform-provider-azurerm/tree/feat/app-gateway-customerrors)

**Comment by tombuildsstuff** [comment]
@mcharriere 

> Hi, I've been working on adding support for Custom errors to the application gateway resource. @tombuildsstuff would you like to add this new feature to the enhancements list?

Rather than extending this issue further, would you mind opening a separate issue for that?

> Should I open a pull request? I have the code in this repo

That'd be great, thanks - we can take a look once it's open :)

Thanks!

**Comment by mcharriere** [comment]
@tombuildsstuff 

> > Hi, I've been working on adding support for Custom errors to the application gateway resource. @tombuildsstuff would you like to add this new feature to the enhancements list?
> 
> Rather than extending this issue further, would you mind opening a separate issue for that?

I've just realized that there is another issue open (#2364) related with custom errors. Thanks!

**Comment by stankevicius** [comment]
Guys, any update on AGW redirect functionality implementation (as well as http2 option) ? Microsoft released it in 2017 and we still do not have it here. It is really critical functionality and quite default in load balancing world. So it would be really nice to have it here and rid out from powershell finally.

**Comment by RustyF** [comment]
You're in luck - PR should be done this week. Just working on some validation. I have no idea how the release cycle works here but I'm sure it won't be that long.

**Comment by RustyF** [comment]
@stankevicius, on the version of code I'm looking at, the http2 option (a property directly against the AGW) , i.e. `enable_http2`, is already there? Or is there a http2 option specifically related to redirection? I can't see that.

**Comment by steve-88** [comment]
> You're in luck - PR should be done this week. Just working on some validation. I have no idea how the release cycle works here but I'm sure it won't be that long.
> 
> Oh, no idea about the http2 bit - I'll have a look.

Finally! Thank you :)

**Comment by stonefury** [comment]
@RustyF Thank you as well!

**Comment by ghostinthewires** [comment]
@stankevicius @RustyF - I have already implemented the HTTP2 Option in this pr https://github.com/terraform-providers/terraform-provider-azurerm/pull/2735 

@stankevicius Can you clarify what you mean by redirection?

**Comment by mjiderhamn** [comment]
@ghostinthewires Search for "redirect" on this issue page and you'll be directed to #552 . Seems that HTTP to HTTPS redirect rules are the primary need.

**Comment by stonefury** [comment]
Correct.  We do a lot of http to https redirection.

**Comment by stankevicius** [comment]
@ghostinthewires 
Yup, I see that http2 option support is really implemented but it is not documented under https://www.terraform.io/docs/providers/azurerm/r/application_gateway.html So I missed it. Thank you for it.

Answer to the second question is http to https redirection between listeners. This is really critical as this rule is going as a default one in almost all cases.

In powershell it looks like as provided below:
` 
provisioner "local-exec" {
        command     = "az network application-gateway redirect-config create --gateway-name ${azurerm_application_gateway.name} -g ${var.resource_group_name} -n http-to-https-redirconfig --type Permanent --include-path true --include-query-string true --target-listener https-listener"
    }

provisioner "local-exec" {
     command     = "az network application-gateway rule create --gateway-name ${azurerm_application_gateway.name} -g ${var.resource_group_name} -n http-to-https-rule --rule-type Basic --http-listener http-listener --redirect-config http-to-https-redirconfig"
}
`

**Comment by ghostinthewires** [comment]
Hi @stankevicius - It is in that document here as was part of my pr: https://www.terraform.io/docs/providers/azurerm/r/application_gateway.html#http2_enabled

OK, I understand your issue now. Thanks for the clarification

**Comment by stankevicius** [comment]
@ghostinthewires thanks in advanced.

By the way, "**http2_enabled**" is exported attribute name only and not the one which we can use in terraform code when we would like enable/disable http2. So real parameter name that we should use is "**enable_http2 = true**" (I find out it from source code) and this one is missing in app gateway parameters usage description in the same page which you provided.

**Comment by ghostinthewires** [comment]
@stankevicius - Good spot, have raised a pr to get this corrected in the docs

**Comment by 1arrow** [comment]
Are we covering ssl_policy here in this enhancement ticket?

https://github.com/terraform-providers/terraform-provider-azurerm/issues/898

I see the above ticket is closed, but there is no references to that feature request here

**Comment by alexeyklots** [comment]
> You're in luck - PR should be done this week. Just working on some validation. I have no idea how the release cycle works here but I'm sure it won't be that long.
> 
> Oh, no idea about the http2 bit - I'll have a look.

Hi @RustyF, as far as I understand you are working on redirect functionality, is is right? When you think it will be ready.
Thanks!

**Comment by Mobe1969** [comment]
Why doesn't "Support for Hostname" cover the ability to set host name on backend_http_settings? The "pick_host_name_from_backend_address" is a totally separate setting, and not the http settings host name. Eg the New-AzureRmApplicationGatewayBackendHttpSettings -HostName parameter

**Comment by mcharriere** [comment]
@Mobe1969 I'm opening a PR with that feature.

**Comment by stonefury** [comment]
@mobe1969. I see what you mean now. Thanks. Would be great to have support for it.

**Comment by jcsturges** [comment]
Huge :+1: for **Support for Redirect Rules** (https://github.com/terraform-providers/terraform-provider-azurerm/issues/552). This one is fairly critical for us; right now we use a workaround via `az network application-gateway` commands but it's not always great.

**Comment by bostonmoto** [comment]
Hello @stankevicius and others.

For those using the above referenced workaround with local-exec and azcli, how do you get around the error below when trying to update other properties gateway in terraform like WAF Config.

 1 error(s) occurred:

* azurerm_application_gateway.application: Error Creating/Updating Application Gateway "appgw" (Resource Group "RGname"): network.ApplicationGatewaysClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="ApplicationGatewayRequestRoutingRuleMustReference" Message="Request Routing Rule /subscriptions/XXXXXXX/resourceGroups/RGname/providers/Microsoft.Network/applicationGateways/appg/requestRoutingRules/rulename must reference a BackendAddressPool." Details=[]

Thanks

**Comment by bostonmoto** [comment]
> For now i have found a workaround, which might be useful for someone else. I have added ignore_changes block to ignore routing rule, so my local-exec part doesnt get overwritten for routing rules needed to have http to https redirection. Same should be working for other block needed.
> 
> resource "azurerm_application_gateway" "aag" {
> name = "${var.prefix}-aag"
> resource_group_name = "${var.rgs_name}"
> location = "${var.region}"
> 
> **lifecycle { ignore_changes = [ "request_routing_rule" ]**
> }

Thanks I tried this also, sadly it seems to work if you run create_plan, however when you run apply_plan is in my case when updating WAF Configuration it throws an error.  Did you try to run an apply after create? see error below.

* azurerm_application_gateway.application: Error Creating/Updating Application Gateway "appgw" (Resource Group "RGname"): network.ApplicationGatewaysClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="ApplicationGatewayRequestRoutingRuleMustReference" Message="Request Routing Rule /subscriptions/XXXXXXXXX7/resourceGroups/UCpscSapp/providers/Microsoft.Network/applicationGateways/appgw/requestRoutingRules/rulename must reference a BackendAddressPool." Details=[]

**Comment by stonefury** [comment]
@bostonmoto - Thanks for the tip. Good idea.

**Comment by bostonmoto** [comment]
@stonefury the workaround does not work. I was actually quoting a response from @dopsy earlier to let him know and wondering if anyone had found something that works.
The issue is now one cannot update AppGateway via terraform if creating this custom redirect rule with azcli, this is really bad if you need to update AppG often to add new sites or are in the middle of a project.

**Comment by stonefury** [comment]
@bostonmoto Thanks for replying back. I jumped the gun there.

**Comment by RustyF** [comment]
With @katbyte and @tombuildsstuff help, redirect rules support is now merged

**Comment by steve-88** [comment]
@RustyF Does this mean we can start using it in our code? What is the next steps?

**Comment by jstewart612** [comment]
@steve-88 Basic source code and project lifecycle management stuff, but tl;dr; this will release with version 1.24.0 of the provider which does not yet have a release date as of this comment.

If you look at the CHANGELOG.md provided in the master branch (now that the PR has been accepted to add the feature), you will see the following note (https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/CHANGELOG.md):

```azurerm_application_gateway - support for redirect rules [GH-2908]```

If you git blame your friend there:  https://github.com/terraform-providers/terraform-provider-azurerm/pull/2908

There's our man @RustyF , bravo!  Cheers for the very important feature addition!

Though a date is not yet set, you're always welcome to follow the Milestones on any Github project, which is usually a good clue on when project maintainers think they're going to be ready to release a given version tag.  1.24.0's appears to reside here (https://github.com/terraform-providers/terraform-provider-azurerm/pulls?q=is%3Aopen+is%3Apr+milestone%3Av1.24.0).

**Comment by SirWayne** [comment]
Hey,

is there a possiblilty to enable the advanced rule configuration for the WAF. I would like to disbale single rules of the ruleset `REQUEST-942-APPLICATION-ATTACK-SQLI`

**Comment by alert101** [comment]
> Are we covering ssl_policy here in this enhancement ticket?
> 
> #898
> 
> I see the above ticket is closed, but there is no references to that feature request here

We are also waiting for ssl_policy. There's another ticket which has been closed in favor of this one.
https://github.com/terraform-providers/terraform-provider-azurerm/issues/451

**Comment by jstewart612** [comment]
@alert101 #1536

**Comment by RustyF** [comment]
@Phydeauxman as I've just added the redirect functionality, I'm keen to help with this.  However, I don't understand the coexistence you're talking about.  You should be able to have path-based redirects but I guess this isn't what you mean?  If you can elaborate in any way, share configs or portal screenshots, that would help. Or is this coexistence with another Gateway feature?

**Comment by Phydeauxman** [comment]
@RustyF there is a lot of confusion when it comes to the different pieces of an App Gateway. On a Request Routing Rule, there are path based maps where you can map traffic patterns to a specific backendpool/httpsetting and then there is redirection where you can send traffic to another listener, or an external URL. I am able to configure a request routing rule via the portal to have both but the code you implemented in the provider does not support this. The provider code is either one or the other.

![image](https://user-images.githubusercontent.com/16844071/55824152-498aab00-5ad1-11e9-85d1-324f6b75cd4f.png)

**Comment by bostonmoto** [comment]
@Phydeauxman do you get an error or what happens when you try to use both those features in the code? The update documention lists 2 optional sections for both redirect rules and path based routing.
https://www.terraform.io/docs/providers/azurerm/r/application_gateway.html

**Comment by Phydeauxman** [comment]
@bostonmoto the error tells you that you can use them together...it is one or the other...which the documentation states. My question was why was it implemented this way when thru the portal, I can configure both to co-exist?

**Comment by RustyF** [comment]
Let's assume the validation is too restrictive and I will aim (this week is the plan) to revise it.  I still need to understand what the two conflicting configs are as I did experience some mutual exclusions when I was testing it. Those "Default.." settings definitely have some conflict with specific rules I'm just not sure which scenario is OK yet.

**Comment by RustyF** [comment]
Hi @Phydeauxman - are you able to send me the fragment of redirect config you've tried to add? I will turn it into a failing test that I can then work against.  Feel free to anonymise it, of course  😉

**Comment by Phydeauxman** [comment]
@RustyF below is the configuration block in my App Gateway that will throw a validation error but can be effected via the portal:

    # Path-based routing rule for UI
    request_routing_rule {
    name                        = "${var.projectPrefix}-${var.les_naming_suffix}-rule"
    rule_type                   = "PathBasedRouting"
    http_listener_name          = "${var.projectPrefix}-${var.les_naming_suffix}-listener"
    backend_address_pool_name   = "${data.terraform_remote_state.apps.lesapp_name}-beap"
    backend_http_settings_name  = "${data.terraform_remote_state.apps.lesapp_name}"
    redirect_configuration_name = "LetsEncryptChallenge"
    url_path_map_name           = "${var.projectPrefix}-${var.les_naming_suffix}"
    }

    url_path_map {
      name = "${var.projectPrefix}-${var.les_naming_suffix}"
      default_backend_address_pool_name  = "${data.terraform_remote_state.apps.lesapp_name}-beap"
      default_backend_http_settings_name = "${data.terraform_remote_state.apps.lesapp_name}"

    path_rule {
      name                       = "${data.terraform_remote_state.apps.lesapiapp_name}"
      paths                      = ["/api*"]
      backend_address_pool_name  = "${data.terraform_remote_state.apps.lesapiapp_name}-beap"
      backend_http_settings_name = "${data.terraform_remote_state.apps.lesapiapp_name}"
      }
    }

This is from the documentation:
`backend_http_settings_name - (Optional) The Name of the Backend HTTP Settings Collection to use for this Path Rule. Cannot be set if redirect_configuration_name is set.`

`redirect_configuration_name - (Optional) The Name of a Redirect Configuration to use for this Path Rule. Cannot be set if backend_address_pool_name or backend_http_settings_name is set.`

**Comment by RustyF** [comment]
Thanks for the snippet @Phydeauxman . Have you tried removing `backend_address_pool_name` and `backend_http_settings_name` from the main `request_routing_rule` config? I think those are not required (and hence the issue).  I believe this is the purpose of the `default...` entries on `url_path_map`.

**Comment by Phydeauxman** [comment]
@RustyF I have not tried that but will.

**Comment by bostonmoto** [comment]
Thanks @Phydeauxman look forward to hearing your updates.

**Comment by Phydeauxman** [comment]
@RustyF @bostonmoto been doing some testing on this today and getting some different results than what I originally got. First off, below is an image of the redirection configuration I was trying to effect:

![image](https://user-images.githubusercontent.com/16844071/56372057-40ce5f00-61cc-11e9-81c0-72f6231df093.png)

The Terraform implementation of the `redirect_configuration` definition does not provide a way to define the `Path` field.

When I start from scratch with no existing App Gateway and I attempt to apply the config below, it gives me no warnings and the apply says it is going to do what I have in the config:

![image](https://user-images.githubusercontent.com/16844071/56373230-868c2700-61ce-11e9-8f1e-c2ba86ac8556.png)

![image](https://user-images.githubusercontent.com/16844071/56373300-a6bbe600-61ce-11e9-9b87-aea859661a89.png)

This apply will fail with the following error:

![image](https://user-images.githubusercontent.com/16844071/56373691-804a7a80-61cf-11e9-84fc-9bebfc3ae8da.png)

If I remove the `include_path` property from the `redirect_configuration` block (property defaults to false), the apply error goes away and the Gateway gets created but it does not create the redirection configuration. My assumption is because a value for the required `Path` field is not present.

**Comment by mcharriere** [comment]
@Phydeauxman There is no Path field in the GO SDK. 

Could you try something like the following?:

```hcl
  request_routing_rule {
    name               = "${var.ui_rule_name}-rule"
    rule_type          = "PathBasedRouting"
    http_listener_name = "${var.projectPrefix}-${var.ui_listener_name}-listener"
    url_path_map_name  = "${var.ui_rule_name}"
  }

  redirect_configuration {
    name          = "LetsEncryptChallenge"
    redirect_type = "Permanent"
    target_url    = "https://lechallenge.blob.core.windows.net/public"
  }

  url_path_map {
    name                               = "${var.ui_rule_name}"
    default_backend_address_pool_name  = "${data.terraform_remote_state.apps.uiapp_name}"
    default_backend_http_settings_name = "${data.terraform_remote_state.apps.uiapp_name}"

    path_rule {
      name                        = "letsencrypt"
      paths                       = ["/.well-known/challenge/*"]
      redirect_configuration_name = "LetsEncryptChallenge"
    }
  }
```
I haven't tested it, but I think that this is the way to achieve what you want.

**Comment by Phydeauxman** [comment]
@mcharriere appears that is not going to work either:

![image](https://user-images.githubusercontent.com/16844071/56389618-024e9980-61f8-11e9-8e33-7c93f134d4ce.png)

**Comment by Phydeauxman** [comment]
@mcharriere I actually went back thru this again and discovered that even though the apply throws that error...it does create the App Gateway with the rule just as I intended it to be. The by product is though...it throws that error, and you can't change or destroy the App Gateway with Terraform now. You have to manually change it or delete it.

**Comment by mcharriere** [comment]
@Phydeauxman Yep, I've tested it myself and I've got it working without that validation placed there. 
It's also odd that all those validations are inside the flatten function. 
I'll test it a bit more and I'll try to send a fix for it.

**Comment by RustyF** [comment]
Thanks for both looking at this - I’m not at my computer at the moment but I couldn’t find that path parameter in the Azure SDK docs either so the GUI is obviously mapping it onto something else. It’s quite possible that the validation isn’t quite right; the docs are a bit lacking in explaining the mutual exclusions. Btw, placing validation in the flatten function appears to be the convention but I don’t have any previous experience to validate that 🤔

**Comment by Phydeauxman** [comment]
@RustyF no worries...it takes a village. Not the first time the SDK or documentation from MS was lacking...nor will it be the last.

**Comment by jeffwmiles** [comment]
[Feature Request 1519](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1519) included a mention of a configuration item "Use for App Service", which would reside on the "backend_http_settings" block. 

This may still need to be implemented; it appears selected in the portal with a big red error (see attached) when I use the following config:
`backend_http_settings {
    name                  = "testext-httpsetting"
    cookie_based_affinity = "Disabled"
    port                  = 80
    protocol              = "HTTP"
    port                  = 443
    protocol              = "HTTPS"
    authentication_certificate = {
      name = "certname.cer"
    }
    request_timeout       = 20
    probe_name = "appgw-siteup-probe"
    pick_host_name_from_backend_address = true
  }`

![2019-05-16 00_08_06-Window](https://user-images.githubusercontent.com/40004712/57830254-c4994e00-776e-11e9-843d-745b9b6b4e1d.png)

Note I don't see the certificate mentioned here, but it does appear if I de-select the "Use for App Service" box. 

A [StackOverflow post ](https://stackoverflow.com/questions/52945056/use-for-app-service-parameter-in-add-azurermapplicationgatewaybackendhttpsetti) references that this option is a combination of using a custom probe and "PickHostNameFromBackendAddress" which is already in my config. 

Is there a configuration property missing, or is Terraform not setting the "PickHostNameFromBackendAddress" properly to flag this value?

**Comment by timja** [comment]
Should this be closed? Most if not all of the issues here are fixed, and would probably be better tracked as individual issues anyway?

**Comment by mbfrahry** [comment]
Hey all. It looks like we got all but one of the issues in this thread so I'll be closing it down. The only one we missed was #1274 which we weren't able to reproduce. If anyone has more information on that please open a new issue with additional information and steps to reproduce.
